bookserver/crud.py

```diff
@@ -118,6 +118,39 @@ async def count_useinfo_for(
         return res.all()
 
 
+async def fetch_page_activity_counts(
+    chapter: str, subchapter: str, base_course: str, course_name: str, username: str
+) -> Dict[str, int]:
+    query = select(Question).where(
+        (Question.subchapter == subchapter)
+        & (Question.chapter == chapter)
+        & (Question.from_source == True)  # noqa: E712
+        & ((Question.optional == False) | (Question.optional == None))
```

Comment:
Nitpick: isn't this `Question.optional != True`?

Reply:
I'm never sure how it will interpret null. this may be a hold over from weird DAL behavior.

Reply:
Probably, since in `models.py` the field is allowed to be null.